### PR TITLE
Updating some RADIUSS packages from LLNL/radiuss-spack-configs

### PR DESCRIPTION
**Pull request by adrienbernede**
## Summary

We have been maintaining a "local" set of packages for a long time. We kept them up-to-date with changes in Spack at a slower pace, while they have improvements that will benefits others.

We are now in sync with Spack, so it’s a good time to submit this PR.

## Changes

- Some packages moved to the CachedCMakePackage class.
- Some packages have newer versions, new patches, conflicts, variants...
- There is a new package (raja-perf).
- Most packages are tuned to work well on LC system, relying on utility functions that are shared instead of duplicated, and ~located in the Camp package (1)~.

~## (1) Handling utility functions (located in Camp package)~

~We are of course open to suggestions regarding the best way to handle those utility functions.~

~Options appears to be:~
~1. Create a Radiuss (or LLNL) utility package.~
~2. Create a Radiuss (or LLNL) build system. ~
~3. Do nothing and submit upstream as-is, waiting for the appropriate capability in Spack.~
4. **Per @tgamblin suggestion: I am moving the utility functions from camp package to CachedCMake build system.**

-> Currently curating the utility functions. Solving conflicts with some logic already merged in https://github.com/spack/spack/pull/37592/files. Testing changes in https://github.com/LLNL/CHAI/pull/237.

### CODE REVIEW

**Comment by adrienbernede** [comment]
@rhornung67 @davidbeckingsale : Big day !

File: var/spack/repos/builtin/packages/caliper/package.py [hunk 1]
```diff
@@ -101,63 +103,94 @@ class Caliper(CMakePackage, CudaPackage, ROCmPackage):
...
 
+    def _get_sys_type(self, spec):
+        sys_type = spec.architecture
+        if "SYS_TYPE" in env:
+            sys_type = env["SYS_TYPE"]
+        return sys_type
+
     def cmake_args(self):
+        options = []
+        return options
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/caliper/package.py]: You could just `return []` unless you're expecting to add arguments here.

File: var/spack/repos/builtin/packages/raja-perf/package.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/raja-perf/package.py]:
```suggestion
# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```

File: var/spack/repos/builtin/packages/raja/package.py [hunk 3]
```diff
@@ -1,12 +1,17 @@
-# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/raja/package.py]:
```suggestion
# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```

File: var/spack/repos/builtin/packages/umpire/package.py [hunk 4]
```diff
@@ -1,15 +1,19 @@
-# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/umpire/package.py]:
```suggestion
# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```

File: var/spack/repos/builtin/packages/camp/package.py [hunk 5]
```diff
@@ -1,21 +1,32 @@
-# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/camp/package.py]:
```suggestion
# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```

File: var/spack/repos/builtin/packages/chai/package.py [hunk 6]
```diff
@@ -1,11 +1,17 @@
-# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/chai/package.py]:
```suggestion
# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```

File: var/spack/repos/builtin/packages/raja-perf/package.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+from spack.package import *
+from .camp import hip_for_radiuss_projects
+from .camp import cuda_for_radiuss_projects
+from .camp import blt_link_helpers
+
+
+class RajaPerf(CachedCMakePackage, CudaPackage, ROCmPackage):
+    """RAJA Performance Suite."""
+
+    homepage = "http://software.llnl.gov/RAJAPerf/"
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/raja-perf/package.py]: I cannot confirm this homepage.  Also, we prefer use of `https` protocol.

**Review by tldahlgren — COMMENTED** [review]
Confirmed all of the branches, tags, and sha256.  Suggestion for the homepage and copyright ending years should be `2023` (or at least not < `2022`).

File: var/spack/repos/builtin/packages/caliper/package.py (hunk 1 above)
Comment (adrienbernede) [reply var/spack/repos/builtin/packages/caliper/package.py]: That, and the fact that cmake_args is defined twice in this package...

File: var/spack/repos/builtin/packages/raja-perf/package.py (hunk 7 above)
Comment (adrienbernede) [reply var/spack/repos/builtin/packages/raja-perf/package.py]: Agreed.

**Comment by adrienbernede** [comment]
@tldahlgren Do we absolutely have to comply to the automated formatting ?
Here is an example of the requested changes that are not for the best IMO:

Screen Shot 2023-08-28 at 3 33 46 PM

File: var/spack/repos/builtin/packages/raja-perf/package.py (hunk 7 above)
Comment (adrienbernede) [reply var/spack/repos/builtin/packages/raja-perf/package.py]: And fixed

File: var/spack/repos/builtin/packages/caliper/package.py (hunk 1 above)
Comment (adrienbernede) [reply var/spack/repos/builtin/packages/caliper/package.py]: Fixed

File: var/spack/repos/builtin/packages/chai/package.py [hunk 8]
```diff
@@ -106,51 +121,73 @@ def cache_name(self):
...
 
     def initconfig_compiler_entries(self):
         spec = self.spec
+        compiler = self.compiler
         entries = super().initconfig_compiler_entries()
-        if "+rocm" in spec:
-            entries.insert(0, cmake_cache_path("CMAKE_CXX_COMPILER", spec["hip"].hipcc))
+
+        # adrienbernede-22-11:
+        #   This was in upstream Spack raja package, but it’s causing the follwing failure Umpire:
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/chai/package.py]:
```suggestion
        #   This was in upstream Spack raja package, but it’s causing the following failure for Umpire:
```

File: var/spack/repos/builtin/packages/raja-perf/package.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+    version("0.11.0", tag="v0.11.0", submodules="True")
+    version("0.10.0", tag="v0.10.0", submodules="True")
+    version("0.9.0", tag="v0.9.0", submodules="True")
+    version("0.8.0", tag="v0.8.0", submodules="True")
+    version("0.7.0", tag="v0.7.0", submodules="True")
+    version("0.6.0", tag="v0.6.0", submodules="True")
+    version("0.5.2", tag="v0.5.2", submodules="True")
+    version("0.5.1", tag="v0.5.1", submodules="True")
+    version("0.5.0", tag="v0.5.0", submodules="True")
+    version("0.4.0", tag="v0.4.0", submodules="True")
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/raja-perf/package.py]: Should've mentioned this before but I didn't dig into the problem.  I get `All fetchers failed` for versions `0.4.0` and `0.5.1`.  

I briefly looked at the site last week and it seemed like they should work.

**Comment by tldahlgren** [comment]
> @tldahlgren Do we absolutely have to comply to the automated formatting ? Here is an example of the requested changes that are not for the best IMO:
> 
> Screen Shot 2023-08-28 at 3 33 46 PM

They are required checks that I do not have the ability to override.  

Someone else complained recently about the value used for the line length but Todd G. pointed out the data he used to determine the value.

File: var/spack/repos/builtin/packages/raja-perf/package.py (hunk 9 above)
Comment (adrienbernede) [reply var/spack/repos/builtin/packages/raja-perf/package.py]: It turns out 0.4.0 and 0.5.1 miss the "v".

I’m fixing this.

Screen Shot 2023-08-28 at 7 05 34 PM

Comment (adrienbernede) [reply var/spack/repos/builtin/packages/raja-perf/package.py]: Fixed

**Comment by adrienbernede** [comment]
> > @tldahlgren Do we absolutely have to comply to the automated formatting ? Here is an example of the requested changes that are not for the best IMO:
> 
> They are required checks that I do not have the ability to override.
> 
> Someone else complained recently about the value used for the line length but Todd G. pointed out the data he used to determine the value.

Here it’s not really about the line length, but rather where to break it.
Code format heuristics don’t have a "coherency" feature, which in this situation leads to much less readable code.

But anyways, I’ll just wait a bit before applying the new format: I am investigating an issue with hip support. And it’s more convenient if I don’t change the packages too much yet.

**Comment by adrienbernede** [comment]
Per @tgamblin suggestion: I am moving the utility functions from camp package to CachedCMake build system.

Ongoing:
- Curating the utility functions.
- Solving conflicts with some logic already merged in https://github.com/spack/spack/pull/37592/files.
- Testing changes in https://github.com/LLNL/CHAI/pull/237.

**Comment by adrienbernede** [comment]
@spack-bot commands

**Comment by adrienbernede** [comment]
@spackbot commands

**Comment by adrienbernede** [comment]
@spackbot fix style

**Comment by adrienbernede** [comment]
@white238 I moved a blt related function to the blt package instead of the cached cmake build-system.
Since it’s very LLNL and BLT specific, it made more sense to me. Thoughts?

**Comment by adrienbernede** [comment]
@tldahlgren Is there a way to get the bot to fix style even in a PR from inside Spack/spack ?

**Comment by alalazo** [comment]
@spackbot fix style

File: var/spack/repos/builtin/packages/camp/package.py [hunk 10]
```diff
@@ -64,33 +53,72 @@ def cmake_args(self):
...
             if not spec.satisfies("cuda_arch=none"):
                 cuda_arch = spec.variants["cuda_arch"].value
                 options.append("-DCMAKE_CUDA_ARCHITECTURES={0}".format(cuda_arch[0]))
                 options.append("-DCUDA_ARCH=sm_{0}".format(cuda_arch[0]))
                 flag = "-arch sm_{0}".format(cuda_arch[0])
                 options.append("-DCMAKE_CUDA_FLAGS:STRING={0}".format(flag))
-        else:
-            options.append("-DENABLE_CUDA=OFF")
 
+        options.append(self.define_from_variant("ENABLE_HIP", "hip"))
```
Comment (adrienbernede) [inline var/spack/repos/builtin/packages/camp/package.py]:
```suggestion
        options.append(self.define_from_variant("ENABLE_HIP", "rocm"))
```

File: var/spack/repos/builtin/packages/camp/package.py [hunk 11]
```diff
@@ -64,33 +53,72 @@ def cmake_args(self):
...
+                if not os.path.exists(mpiexec):
+                    mpiexec = os.path.join(spec["mpi"].prefix.bin, "mpiexec")
+
+            if not os.path.exists(mpiexec):
+                msg = "Unable to determine MPIEXEC, %s tests may fail" % self.pkg.name
+                tty.warn(msg)
+            else:
+                # starting with cmake 3.10, FindMPI expects MPIEXEC_EXECUTABLE
+                # vs the older versions which expect MPIEXEC
+                if self.pkg.spec["cmake"].satisfies("@3.10:"):
```
Comment (adrienbernede) [inline var/spack/repos/builtin/packages/camp/package.py]:
```suggestion
                if spec["cmake"].satisfies("@3.10:"):
```

File: var/spack/repos/builtin/packages/camp/package.py [hunk 12]
```diff
@@ -64,33 +53,72 @@ def cmake_args(self):
...
+                    mpiexec = "/usr/bin/srun"
+                else:
+                    mpiexec = os.path.join(spec["slurm"].prefix.bin, "srun")
+            else:
+                mpiexec = os.path.join(spec["mpi"].prefix.bin, "mpirun")
+                if not os.path.exists(mpiexec):
+                    mpiexec = os.path.join(spec["mpi"].prefix.bin, "mpiexec")
+
+            if not os.path.exists(mpiexec):
+                msg = "Unable to determine MPIEXEC, %s tests may fail" % self.pkg.name
```
Comment (adrienbernede) [inline var/spack/repos/builtin/packages/camp/package.py]:
```suggestion
                msg = "Unable to determine MPIEXEC, %s tests may fail" % self.name
```

**Comment by jhdavis8** [comment]
I'm not sure what state this update is in but I get the following error when trying to build `raja+rocm~examples~exercises amdgpu_target=gfx90a generator=ninja %gcc@11.2.0` on OLCF Frontier using this branch:
```
==> Installing raja-2023.06.1-nqu7y2mnoh5qi5sksrbgdc4ucgqlpa6y [29/30]
==> No binary for raja-2023.06.1-nqu7y2mnoh5qi5sksrbgdc4ucgqlpa6y found: installing from source
==> Using cached archive: /autofs/nccs-svm1_home1/jhdavis/spack/var/spack/cache/_source-cache/git//LLNL/RAJA.git/9b5f61edf3aa1e6fdbc9a4b30828c81504639963.tar.gz
==> Warning: Fetching from mirror without a checksum!
  This package is normally checked out from a version control system, but it has been archived on a spack mirror.  This means we cannot know a checksum for the tarball in advance. Be sure that your connection to this mirror is secure!
==> No patches needed for raja
==> raja: Executing phase: 'initconfig'
==> Error: AttributeError: 'Adapter' object has no attribute 'compiler'

The 'raja' package cannot find an attribute while trying to build from sources. This might be due to a change in Spack's package format to support multiple build-systems for a single package. You can fix this by updating the build recipe, and you can also report the issue as a bug. More information at https://spack.readthedocs.io/en/latest/packaging_guide.html#installation-procedure

/autofs/nccs-svm1_home1/jhdavis/spack/lib/spack/spack/build_systems/cached_cmake.py:293, in initconfig_hardware_entries:
        290            hip_link_flags = ""
        291            if "%gcc" in spec or spec_uses_toolchain(spec):
        292                if "%gcc" in spec:
  >>    293                    gcc_bin = os.path.dirname(self.compiler.cxx)
        294                    gcc_prefix = os.path.join(gcc_bin, "..")
        295                else:
        296                    gcc_prefix = spec_uses_toolchain(spec)[0]
```
(edited to fix mistake in output snippet)

**Comment by adrienbernede** [comment]
@jhdavis8 thank you for testing this branch and providing feedback.

There was indeed an error we never spotted in our CI due to not covering that case. Great that you found it !

**Comment by adrienbernede** [comment]
@spackbot commands

**Comment by adrienbernede** [comment]
@spackbot rebuild everything

**Comment by adrienbernede** [comment]
@spackbot rebuild everything

**Comment by adrienbernede** [comment]
@spackbot run pipeline

**Comment by adrienbernede** [comment]
@spackbot rebuild everything

**Comment by adrienbernede** [comment]
This is deprecated. I’ll create a new PR when I push missing changes to the RADIUSS packages.

**Pull request closed without merging**